snowshu/core/graph.py

```diff
@@ -49,26 +48,17 @@ def catalog_difference(source_graph: Union["SnowShuGraph", networkx.Graph],
         if isinstance(source_graph, SnowShuGraph):
             source_graph = source_graph.graph
 
-        non_isolated_relations = {}
-        for relation, relation_base in source_graph.edges:
-            if non_isolated_relations.get(relation_base):
-                non_isolated_relations[relation_base].append(relation)
-            else:
-                non_isolated_relations[relation_base] = [relation]
-
-        non_isolated_relations = [(relation_base, *relations)
-                                  for relation_base, relations in non_isolated_relations.items()]
-        isolated_relations = list(networkx.isolates(source_graph))
-
-        non_isolated_relations_difference = chain.from_iterable([entries for entries in non_isolated_relations
-                                                                 if not all(entry in target_catalog
-                                                                            for entry in entries)])
+        source_nodes = set(source_graph.nodes)
+        # relations not found in target catalog
+        missing_relations = source_nodes.difference(target_catalog)
+        # generate actual list since source_graph will be modified
+        weakly_connected_subgraphs = [comp for comp in networkx.weakly_connected_components(source_graph)]  # noqa pylint: disable=unnecessary-comprehension
 
-        isolated_relations_difference = set(isolated_relations).difference(target_catalog)
+        # if a component has no missing relations, remove those nodes from the graph
+        for component in weakly_connected_subgraphs:
+            if not missing_relations.intersection(component):
```

Comment:
btw, should we bump up the networkx? the actual one is 2.8.7 while we have 2.6.2

Reply:
Probably worth updating most pins at some point. I think we only updated the pins recently for running snowshu on arm machines.

I'll create a ticket for doing that dependency update (along with dependency updates in other repos)